### PR TITLE
hierarchical clustering

### PR DESCRIPTION
**Pull request by pius-ng3a**
---

Checklist:

- [ ] I have refactored my code out from `notebooks/`
- [ ] I have checked the code runs
- [ ] I have tested the code
- [ ] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [ ] I have documented the code
  - [ ] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [ ] I have explained the feature in this PR or (better) in `output/reports/`
- [ ] I have requested a code review

### CODE REVIEW

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+# %%
+thresh = 1.2
+clusters = hcluster.fclusterdata(comments, thresh, criterion="distance")
+
+
+# %%
+plt.figure(figsize=(10, 7))
+plt.title("Dendrograms")
+dend = hcluster.dendrogram(hcluster.linkage(comments, method="ward"))
+plt.axhline(y=6, color="r", linestyle="--")
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py:132]: This is interesting. Looks to me like 5 prominent clusters. Can you apply these to the data and see what the comments are in each cluster? It would be interesting to see if it has picket up any meaningful groups that are maybe more granular than the three codes.

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+# %%
+# now apply agglomerative clustering to the comments using the determined number of clusters.
+cluster = AgglomerativeClustering(n_clusters=3, affinity="euclidean", linkage="ward")
+cluster.fit_predict(comments)
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py:137]: The next step here is to see what number of these predicts are the same as the actual labels. So you'll need to look at the comments in each predicted cluster and see which label (if any) they are trying to represent. Then maybe look at the correlation between the predicted and the actual labels? You basically wanted to know, for actual label 1, what percentage are predicted cluster 1, what % cluster 2 ... ect

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+# %%
+from sklearn.datasets import make_blobs
+from sklearn.cluster import DBSCAN
+
+# %%
+stop_words = get_stop_words("fr")
+
+# %%
+tfidf_vectorizer = TfidfVectorizer(
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py:115]: I was speaking about to George about this step and he suggested using a sentence transformer to create embeddings that can go into the clustering algorithm. As he said information can be lost with tf-idf which is a bit more basic.

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+# %%
+data.code.factorize()[0]
+
+# %% [markdown]
+# # Concern
+# How does one determine the number of clusters in hierarchical clustering and proceed to clustering without human intervention?
+
+# %%
+# trying meanshift algorithm
+ms = MeanShift()
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py:148]: This looks like a good way to identify the number of clusters. Interesting it only sees 1. I wonder whether the size of the dataset has an affect here. Might be good to test on all comments in rumours, beliefs and observations to see if this gives more than 1. I'd also be interested to see if using the sentence tranformer instead of tfi-idf has an affect.

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+# DBSCAN clustering of the comments
+distance = [0.2, 0.5, 1, 1.2, 1.5, 1.7, 2, 3, 5]
+label_number = []
+for epsi in distance:
+    db = DBSCAN(eps=epsi, min_samples=50).fit(comments)
+    label_number.append(len(np.unique(db.labels_)))
+print(label_number)
+
+# %%
+labels
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py:168]: A bit confused. Is this the result of using db-scan or still the result from using MeanShift?

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+cluster_centers = ms.cluster_centers_
+n_clusters_ = len(np.unique(labels))
+print("Number of estimated clusters:", n_clusters_)
+
+# %%
+labels
+
+# %%
+# DBSCAN clustering of the comments
+distance = [0.2, 0.5, 1, 1.2, 1.5, 1.7, 2, 3, 5]
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py:160]: Can you explain what the distance variable is doing here?

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+label_number = []
+for epsi in distance:
+    db = DBSCAN(eps=epsi, min_samples=50).fit(comments)
+    label_number.append(len(np.unique(db.labels_)))
+print(label_number)
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py:165]: Is this an output from the clustering of the distance variable or the actual comments?

**Review by izzyStewart — COMMENTED** [review]
Thanks Pius looks like you've done some really good investigative work. Have left some comments. A few general things:

- I think it might be useful to try using a sentence transformer model like you did with the rumour matching work and see if that yields better results than tf-idf. I am planning to try this with the classification model also.
- Its a good point you made about how we approach the task without knowing the number of clusters at the start. Does DB-Scan allow for this? I left some comments in that section where I was a little unclear about the process.

When I spoke to George last week he said he thinks this process could be done by clustering AND/OR using semantic search (like you did with your rumour matching work). I am still unsure how the end-to-end process could look with semantic search so I think the meeting on Wednesday will help there. Maybe have a think if you can see a process that could work there? 

One thought I had was if you could use semantic search on the pool of unlabelled comments as a new comment comes in where you could loop through different values for k and see if good cluster comes out of it with the new comment. This would mean clusters are only attempted on new incoming comments but I think that would be ok as the idea is when the model is running, new comments will trickle in and so comments that can't be labelled will potentially appear one at a time. 

I have put my thoughts together on how some possible workflows could look: https://docs.google.com/document/d/16tjSEb5RpaSU3ER1fnBOkR2WjB5_dXTJBPQLbP8ftRo/edit?usp=sharing
Note: Steps 1-2 apply for both methods
